ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/encoders/TomlAbstractEncoder.kt

```diff
@@ -0,0 +1,388 @@
+     *
+     * @param key
+     */
+    protected fun setKey(key: String) {
+        attributes.key = when {
+            key matches bareKeyRegex -> key
+            key matches literalKeyCandidateRegex -> "'$key'"
+            else -> "\"$key\""
+        }
+    }
+
+    /**
+     * Creates a new array encoder instance, with this encoder as its parent.
+     *
+     * @param rootNode The new encoder's root node.
+     * @param attributes The new encoder's attributes.
+     * @return The new instance.
+     */
+    protected open fun arrayEncoder(
+        rootNode: TomlNode,
+        attributes: Attributes = this.attributes.child()
+    ): TomlAbstractEncoder =
+            TomlArrayEncoder(
+                rootNode,
+                parent = this,
+                elementIndex,
+                attributes,
+                inputConfig,
+                outputConfig,
+                serializersModule
+            )
+
+    /**
+     * Creates a new inline table encoder instance, with the encoder as its parent.
+     *
+     * @param rootNode The new encoder's root node.
+     * @return The new instance.
+     */
+    protected open fun inlineTableEncoder(rootNode: TomlNode): TomlAbstractEncoder =
+            TomlInlineTableEncoder(
+                rootNode,
+                parent = this,
+                elementIndex,
+                attributes.child(),
+                inputConfig,
+                outputConfig,
+                serializersModule
+            )
+
+    /**
+     * Creates a new table encoder instance.
+     *
+     * @param rootNode The new encoder's root node.
+     * @return The new instance.
+     */
+    protected open fun tableEncoder(rootNode: TomlNode): TomlAbstractEncoder =
+            TomlMainEncoder(
+                rootNode,
+                elementIndex,
+                attributes.child(),
+                inputConfig,
+                outputConfig,
+                serializersModule
+            )
+
+    /**
+     * @property parent The parent to inherit default values from.
+     * @property key The current element's key.
+     * @property isMultiline Marks subsequent key-string or array pair elements to
+     * be written as multiline.
+     * @property isLiteral Marks subsequent key-string pair elements to be written
+     * as string literals.
+     * @property intRepresentation Changes how subsequent key-integer pair elements
+     * are represented.
+     * @property isInline Marks subsequent table-like elements as inline. Tables
+     * will be written as inline tables.
+     * @property comments Comment lines to be prepended before the next element.
+     * @property inlineComment A comment to be appended to the end of the next
+     * element's line.
+     * @property isImplicit Whether the current property is implicitly defined in
+     * its child, i.e. the table `[a]` in `[a.b]`.
+     */
+    public data class Attributes(
```

Comment:
why do we need this class? also - can we move it to a separate file, I see that it used in several places

Reply:
It just makes keeping track of and passing element information easier. Sure, I'll move it to a separate file

Reply:
Got it, ok, let’s move to a separate class, right now this one looks overloaded